torch/csrc/distributed/c10d/ProcessGroupNCCL.cpp

```diff
@@ -490,53 +509,34 @@ void ProcessGroupNCCL::WorkNCCL::synchronizeInternal(
 
   // In case of blocking, wait for the operation to complete.
   if (blockingWait_) {
-    // Wait for the operation to complete.
     while (!isCompleted()) {
-      if (timedOut()) {
-        // When operation times out due to some errors that are not
-        // detected by nccl communicators, ncclCommWatchdog can not check this
-        // time out error and thus can not abort ncclComms accordingly.
-        // So explicitly abort ncclComms here before throwing this timed out
-        // exception to users, after this, ncclCommWatchdog can detect nccl
-        // communicators are aborted and clean up devNCCLCommMap_ accordingly.
-        // if throwing timed out exception without aborting nccl communicators
-        // here, it was observed that CUDA GPU will have 100% utilization and
-        // can not run new events successfully.
-
-        std::stringstream ss;
-        ss << *this;
-        auto timeoutErrorMsg =
-            c10::str("Work ", ss.str(), " timed out in call to wait().");
-        for (const auto& ncclComm : ncclComms_) {
-          ncclComm->ncclCommAbort(timeoutErrorMsg);
-          const auto& storeKey = getNcclAbortedCommStoreKey(
-              buildNcclUniqueIdStr(ncclComm->getNcclId()));
-          auto rankStr = std::to_string(rank_);
-          store_->set(storeKey, rankStr);
-          LOG(INFO) << "[Rank " << rank_
-                    << "] Wrote aborted communicator id to store: " << storeKey;
-        }
-        auto currentTimepoint = std::chrono::steady_clock::now();
-        auto timeElapsed =
-            std::chrono::duration_cast<std::chrono::milliseconds>(
-                currentTimepoint - workStartTime_);
+      bool timedOut = checkTimeout();
+      // Explicitly abort ncclComms here before throwing this timed out
+      // exception to users.
+      // If throwing timed out excepiton without aborting nccl communicators
+      // here, it was observed that CUDA GPU will have 100% utilization and
+      // can not run new events successfully.
+      if (timedOut) {
         std::string exceptionMsg = c10::str(
             "[Rank ",
             rank_,
-            "] ",
-            "Caught collective operation timeout: ",
+            "] Work ",
             (*this),
-            " ran for ",
-            timeElapsed.count(),
-            " milliseconds before timing out.");
-        TORCH_CHECK(false, exceptionMsg);
+            " timed out in blocking wait (NCCL_BLOCKING_WAIT=1).");
+        LOG(ERROR) << exceptionMsg;
+        break;
       }
-      // Check for errors and throw appropriate exception.
-      checkAndThrowException();
+      // Yield
       std::this_thread::sleep_for(
           std::chrono::milliseconds(kSynchronizeBusyWaitMillis));
     }
-    checkAndThrowException();
+    // exception() includes timeout and error during blocking wait
+    if (exception()) {
+      // Abort NCCL communicators
+      abort();
+      // Throw exception (from main thread here)
```

Comment:
does this mean watchdog errors are thrown from the main thread now?

Reply:
In blocking wait mode, yes. (Because the main thread is already blocking waiting.)

And I think that is already the case before, see:
https://github.com/pytorch/pytorch/pull/97066/files#diff-c4e79e1dee3eea72b89d4471b91de8a7dae284e27bd5c32ed3e5296a561e4ee9L539

The change here just de-duplicates `checkAndThrowException()` and `handleException()`. Hope that make sense.